hw_i2c/sample-implementations/cypress_modus_toolbox/sensirion_hw_i2c_implementation.c

```diff
@@ -0,0 +1,231 @@
+ * returned.
+ *
+ * @param address 7-bit I2C address to read from
+ * @param data    pointer to the buffer where the data is to be stored
+ * @param count   number of bytes to read from I2C and store in the buffer
+ * @returns 0 on success, error code otherwise
+ */
+int8_t sensirion_i2c_read(uint8_t address, uint8_t* data, uint16_t count)
+{
+	uint8_t status = STATUS_FAIL;
+	cy_en_scb_i2c_status_t errorStatus;
+	uint32_t masterStatus;
+	/* Timeout 1 sec (one unit is us) */
+	uint32_t timeout = 1000000UL;
+
+	/* Setup transfer specific parameters */
+	cy_stc_scb_i2c_master_xfer_config_t transferConfig =
+	{
+		.slaveAddress = address,
+		.buffer       = NULL,
+		.bufferSize   = 0U,
+		.xferPending  = false
+	};
+	transferConfig.buffer     = data;
+	transferConfig.bufferSize = count;
+
+	/* Initiate read transaction */
+	errorStatus = Cy_SCB_I2C_MasterRead(mI2C_HW, &transferConfig, &mI2C_context);
+	if(errorStatus == CY_SCB_I2C_SUCCESS)
+	{
+		/* Wait until master complete read transfer or time out has occured */
+		do
+		{
+			masterStatus  = Cy_SCB_I2C_MasterGetStatus(mI2C_HW, &mI2C_context);
+			Cy_SysLib_DelayUs(CY_SCB_WAIT_1_UNIT);
+			timeout--;
+
+		} while ((0UL != (masterStatus & CY_SCB_I2C_MASTER_BUSY)) && (timeout > 0));
+
+	   if (timeout <= 0)
+		{
+			/* Timeout recovery */
+			Cy_SCB_I2C_Disable(mI2C_HW, &mI2C_context);
+			Cy_SCB_I2C_Enable (mI2C_HW);
+		}
+		else
+		{
+			/* Check transfer status */
+			if (0u == (MASTER_ERROR_MASK & masterStatus))
+			{
+				status = STATUS_OK;
+			}
+		}
+	}
+	return (status);
+}
+
+/**
+ * Execute one write transaction on the I2C bus, sending a given number of bytes.
+ * The bytes in the supplied buffer must be sent to the given address. If the
+ * slave device does not acknowledge any of the bytes, an error shall be
+ * returned.
+ *
+ * @param address 7-bit I2C address to write to
+ * @param data    pointer to the buffer containing the data to write
+ * @param count   number of bytes to read from the buffer and send over I2C
+ * @returns 0 on success, error code otherwise
+ */
+int8_t sensirion_i2c_write(uint8_t address, const uint8_t* data, uint16_t count)
+{
+	uint8_t status = STATUS_FAIL;
+	cy_en_scb_i2c_status_t  errorStatus;
+	uint32_t masterStatus;
+	/* Timeout 1 sec (one unit is us) */
+	uint32_t timeout = 1000000UL;
+
+	/* Setup transfer specific parameters. */
+	cy_stc_scb_i2c_master_xfer_config_t transferConfig =
+	{
+	    .slaveAddress = address,
+	    .buffer       = NULL,
+	    .bufferSize   = 0U,
+	    .xferPending  = false
+	};
+	transferConfig.buffer     = data;
+	transferConfig.bufferSize = count;
+
+	/* Initiate write transaction. */
+	errorStatus = Cy_SCB_I2C_MasterWrite(mI2C_HW, &transferConfig, &mI2C_context);
+	if(errorStatus == CY_SCB_I2C_SUCCESS)
+	{
+		/* Wait until master complete read transfer or time out has occurred */
+		do
+		{
+			masterStatus  = Cy_SCB_I2C_MasterGetStatus(mI2C_HW, &mI2C_context);
+			Cy_SysLib_DelayUs(CY_SCB_WAIT_1_UNIT);
```

Comment:
The `CY_SCB_WAIT_1_UNIT` is really just 1us? Did you check that the timeout kicks in after ~1s?